### ISSUE TITLE
CPU_Delta > 3 000 000 ms, and tasks = 1

### ISSUE THREAD

**Issue by rrg92** [amachanic/sp_whoisactive#2]
How this is possible ?

![image](https://user-images.githubusercontent.com/16887793/62013480-94f0c400-b169-11e9-8ef1-4327639c3c82.png)

This resultset is from a run using @delta_interval  = 1 ... With just one task, in on second, how is possible get more than 1000ms in delta?

Context: I ran it (latest verison, 11.32) on a very busy CPU system (24 cores, virtual machine).

**Comment by amachanic** [comment]
What version of SQL Server? The slightly-off numbers (at least in the 1,000 range) are easy to rationalize based on timing. Obviously 3MM milliseconds in a second would be a bit tougher to swallow. Will need a bit more info, though, before I can tell whether it's a bug in sp_whoisactive or a DMV issue.

**Comment by jzabroski** [comment]
@amachanic Welcome to GitHub.  I recommend setting up Issue Templates so people fill in things like SQL Server Version, possibly using Glen Barry's version details script to get OS details (always tricky due to VM host lying about guest OS details), etc

**Comment by amachanic** [comment]
Thanks @jzabroski. I'll get to work on that.

**Comment by rrg92** [comment]
> What version of SQL Server? The slightly-off numbers (at least in the 1,000 range) are easy to rationalize based on timing. Obviously 3MM milliseconds in a second would be a bit tougher to swallow. Will need a bit more info, though, before I can tell whether it's a bug in sp_whoisactive or a DMV issue.

I remember to see this behavior in most recent versions, like 2014 and 2016. Take a look on this recent collect from an Azure Managed Instance:

![image](https://user-images.githubusercontent.com/16887793/63220210-192adb80-c159-11e9-84a0-bb3c8f922699.png)

The delta interval is just 1 second with 8 CPUs... So, it is conceptually impossible this CPU deltas!
I noted that whenever this occurs, the SQL is using high CPU, with parallel queries!

Maybe this issue is not with whoisactive, and with some DMV that you use to calculate the CPU difference (i don't look the code yet, but i guess that you query from sys.dm_os_workers and sys.dm_os_threads, right!?)

Sorry for long time to answer, and thanks for your attention

**Comment by amachanic** [comment]
Correct, sys.dm_os_threads/workers. I think however that my complex workaround is not necessary in new versions of SQL Server, and that the correct thing to do here is to not delta this way any longer.

At some point the DMVs got better at updating CPU time even in parallel queries. Around SQL Server 2016, I think. I recall that there was a blog post by Pedro Lopes on the topic, but I can't find it at the moment.

Is anyone able to share the relevant info so that a version check can be added?

**Comment by rrg92** [comment]
Nice. ANother suggestion is add a parameter (e.g: @cpu_delta_source), where the user can choose if want see delta from this low level dmv or from sys.dm_exec_requests...

**Comment by jzabroski** [comment]
@copilot find this

"recall that there was a blog post by Pedro Lopes on the topic, but I can't find it at the moment."

**Comment by amachanic** [comment]
> can copilot try that. I can fork and feed the prompt

You're welcome to give it a shot. I'm not hopeful that it will work 🤷

**Comment by erikdarlingdata** [comment]
Let's pit CoPilot and Claude against each other and see who does better.

**Comment by jzabroski** [comment]
I was testing to see if it could inferred that. I gave it the laziest prompt possible. I said fix amachanic/sp_whoisactive/2 lol

**Comment by jzabroski** [comment]
Gemini found the issue: https://support.microsoft.com/en-us/topic/kb2940348-fix-worker-time-in-the-dmvs-sys-dm-exec-procedure-stats-and-sys-dm-exec-query-stats-for-natively-compiled-stored-procedures-is-reported-incorrectly-in-sql-server-2014-44220a30-14ce-d910-2d22-508e5784c568#:~:text=Resolution,cumulative%20update%20of%20SQL%20Server.
